rrdd/xcp_rrdd.ml

```diff
@@ -538,6 +538,121 @@ let monitor_loop () =
+ *
+ * To help plugins with this, we ignore RRD files with a *.tmp suffix.
+ * This gives a plugin the possibility to use an atomic rename(2) call.
+ *)
+
+module type DISCOVER = sig
+	val start: unit -> Stdext.Threadext.Thread.t
+end
+
+module Discover: DISCOVER = struct
+	let directory = Rrdd_server.Plugin.base_path
+
+	(** [is_valid f] is true, if [f] is a filename for an RRD file.
+	 *  Currently we only ignore *.tmp files *)
+	let is_valid file =
+		not @@ Filename.check_suffix file ".tmp"
+
+	let events_as_string
+		: Inotify.event_kind list -> string
+		= fun es ->
+		es
+		|> List.map Inotify.string_of_event_kind
+		|> String.concat ","
+
+	(* [register file] is called when we found a new file in the watched
+	 * directory. We do not verify that this is a proper RRD file.
+	 * [file] is not a complete path but just the basename of the file.
+	 * This corresponds to how the file is used by the Plugin module,
+	 * *)
+	let register file =
+		info "RRD plugin %s discovered - registering it" file;
+		let info  = Rrd.Five_Seconds in
+		let v2    = Rrd_interface.V2 in
+		Rrdd_server.Plugin.Local.register () ~uid:file ~info ~protocol:v2
+		|> ignore (* seconds until next reading phase *)
+
+	(* [deregister file] is called when a file is removed from the watched
+	 * directory *)
+	let deregister file =
+		info "RRD plugin - de-registering %s" file;
+		Rrdd_server.Plugin.Local.deregister () ~uid:file
+
+	(* Here we dispatch over all events that we receive. Note that
+	 * [Inotify.read] blocks until an event becomes available. Hence, this
+	 * code needs to run in its own thread. *)
+	let watch dir =
+		let fd          = Inotify.create () in
+		let selectors   =
+			[ Inotify.S_Create
+			; Inotify.S_Delete
+			; Inotify.S_Moved_to
+			; Inotify.S_Moved_from
+			] in
+		let rec loop = function
+			| []  -> Inotify.read fd |> loop
+			| (_, [Inotify.Create], _, Some file)::es when is_valid file ->
+				( register file (* only basename *)
+				; loop es
+				)
+			| (_, [Inotify.Delete], _, Some file)::es when is_valid file ->
+				( deregister file (* only basename *)
+				; loop es
+				)
+			| (_, [Inotify.Moved_to], _, Some file)::es when is_valid file ->
+				( register file (* only basename *)
+				; loop es
+				)
+			| (_, [Inotify.Moved_from], _, Some file)::es when is_valid file ->
+				( deregister file (* only basename *)
+				; loop es
+				)
+			| (_, events, _, None)::es ->
+				( debug "RRD plugin discovery - ignoring %s" (events_as_string events)
+				; loop es
+				)
+			| (_, events, _, Some file)::es ->
+				( debug  "RRD plugin discovery - ignoring  %s: %s"
+					file (events_as_string events)
+				; loop es
+				)
+		in
+			( Inotify.add_watch fd dir selectors |> ignore
+			; loop []
+			)
+
+	(* [scan] scans a directory for plugins and registers them *)
+	let scan dir =
+		debug "RRD plugin - scanning %s" dir;
+		Sys.readdir dir
+		|> Array.to_list
+		|> List.iter register
```

Comment:
Yes - I should add it. The expectation is that these tmp files are very short lived and don't exist at startup time. But they could exist when we recover from a crash. In any case, we should always ignore tmp files.